[BLOG] Added blog post for enhancing developer experience at SciPy - Intel oneAPI, MSVC and spin support (#893)
Quansight/Quansight-website: apps/labs/posts/intel-oneapi-msvc-spin-scipy.md

---
title: 'Enhancing Developer Experience at SciPy - Intel oneAPI/MSVC Support and Migrating to spin'

> published: February 21, 2025

Reviewer: Just a note to ensure the date is updated before merging this PR.

authors: [gagandeep-singh]
description: 'Highlights the work done to improve developer experience at SciPy, specifically on supporting Intel oneAPI/MSVC and spin'
category: [Developer workflows]
featuredImage:
  src: /posts/intel-oneapi-msvc-spin-scipy/scipy_logo_img_featured.png
  alt: 'SciPy logo'
hero:
  imageSrc: /posts/intel-oneapi-msvc-spin-scipy/scipy_logo_img_hero.png
  imageAlt: 'SciPy logo'
---

> In this blog post, we’ll discuss the improvements we made to enhance the developer experience at SciPy. Our work focused on two key areas:

Reviewer: ```suggestion
Over the past year, we worked on enhancing the development experience of SciPy. We focused on two key areas:
```

The start feels a little abrupt, hence this suggestion for a softer introduction.

1. Intel oneAPI and MSVC Support in SciPy
2. Transitioning from `dev.py` to `spin`

First, let’s explore the motivation behind these updates.

Intel provides a comprehensive toolkit for high-performance computing (HPC) and numerical computing use cases, known as [oneAPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/overview.html).

> Given that SciPy is heavily used in HPC and numerical computing, supporting the ability to build SciPy with Intel oneAPI was a natural and important step.
> By doing so, we aim to improve the experience for SciPy users who rely on Intel oneAPI as their primary toolkit.

Reviewer: ```suggestion
SciPy is heavily used in HPC and numerical computing.
Adding support for SciPy builds with Intel oneAPI was a natural and important step to improve the experience for SciPy users who rely on Intel oneAPI as their primary toolkit.
```

Additionally, since MSVC is the de facto compiler for C/C++ on Windows, we added continuous integration (CI) jobs to test SciPy compilation with MSVC, further expanding the compatibility of SciPy across platforms.

Now, let’s talk about the transition from `dev.py` to `spin`.

> `dev.py` (also known as `do.py` [[1]](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was a significant step forward in improving the development workflow.

Reviewer: ```suggestion
`dev.py` (also known as `do.py`, learn more in [The evolution of the SciPy developer CLI](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was a significant step forward in improving the development workflow.
```

We use this paradigm for footnotes. For links, having descriptive text is more accessible.

However, `spin` builds on `dev.py`, taking inspiration from its design while generalizing the interface to work with a wider range of projects without the need for heavy customization.
As a result, SciPy will directly benefit from the improvements made in `spin` and vice versa.
Moreover, `spin` offers an additional benefit: contributors familiar with `numpy` will find it easier to navigate, thanks to the shared foundation and streamlined approach.

Through these updates, we aim to provide a smoother and more flexible development experience for both current and new contributors to SciPy.

#### Intel oneAPI Support in SciPy

Intel offers oneAPI, a comprehensive toolkit designed for high-performance computing (HPC) applications.

> It includes essential components such as `icx` (for C), `icpx` (for C++), `ifx` (for Fortran), and MKL (Math Kernel Library), which provides optimized mathematical routines like BLAS, LAPACK, and fast FFTs [[2]](https://en.wikipedia.org/wiki/Math_Kernel_Library).

Reviewer: ```suggestion
It includes essential components such as `icx` (for C), `icpx` (for C++), `ifx` (for Fortran), and [MKL (Math Kernel Library)](https://en.wikipedia.org/wiki/Math_Kernel_Library), which provides optimized mathematical routines like BLAS, LAPACK, and fast FFTs.
```

This issue also references [gh-20728](https://github.com/scipy/scipy/issues/20728), where the author is building SciPy using `icx`, `icpx`, `ifx`, and MKL, leveraging the full Intel oneAPI toolkit.
Additionally, the failure to build SciPy with MSVC [[gh-20860]](https://github.com/scipy/scipy/issues/20860) pointed to specific build challenges with MSVC.
We began by addressing these two issues.

One notable challenge with MSVC is its limited support for handling runtime-sized arrays on the stack.
For example, expressions like `std::complex<double> cwrk[n];` fail to compile unless `n` is a constant.
To work around this, we opted for dynamic memory allocation on the heap, using `new`, ensuring proper memory deallocation with `delete` at the end.
While this approach is manual, it remains the only viable solution with MSVC.
Interestingly, Clang (due to its LLVM backend) does not have this limitation, making it more flexible in this regard.

Once we addressed this issue, we added a CI job to test SciPy's compilation with MSVC + `ifx` + OpenBLAS, which helped prevent future regressions in the MSVC build.
This also marked an early step toward supporting Intel oneAPI within SciPy.

> The next major hurdle involved the [`arpack`](https://github.com/opencollab/arpack-ng) issue, which required full Intel oneAPI support to resolve.

Reviewer: I don't think I understand what the "arpack issue" is? Could we link to any issue on the scipy tracker that has more context?

Author: I am referring to https://github.com/scipy/scipy/issues/20728. You are right, just linking it here as well should make it clear. I have referred to this issue in lines 42-44.

Author: Done. Let me know if it looks good to you.

On Linux, the issue was already fixed. However, there were still some challenges when using `ifx`, such as the failure of the `test_equal_bounds` test in `scipy/optimize/tests/test_optimize.py` due to floating-point discrepancies.
To address this, we increased tolerances for several tests to accommodate the slightly reduced accuracy/precision when using `ifx`.
This was all handled in the PR, `BUG/CI: Compile and run tests with ifx + MKL on Linux` [[gh-21173]](https://github.com/scipy/scipy/pull/21173), where I also added a CI job to test the combination of `gcc` + `g++` + `ifx` + MKL on Linux, marking another important step toward full Intel oneAPI support for SciPy.

The final milestone for Intel oneAPI support was the CI job for testing SciPy with `icx`, `icpx`, `ifx`, and MKL [[gh-21254]](https://github.com/scipy/scipy/pull/21254).

> In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux. However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`. Despite several attempts to fix the issue, all solutions led to dead ends. As a result, we decided to pause the effort on Windows for now, as the work required did not seem to justify the outcome.

Reviewer: ```suggestion
In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux.

However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`. Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work, we decided to pause the effort on Windows and focus on on more scoped tasks.
```

3. **Installation**: Finally, `meson install` installs the compiled libraries and source files (for Python) into the installation directory.

With this context in place, let's dive into the specifics of the work we did.

We decided to transition from `dev.py` to `spin` due to a series of issues reported with `dev.py`. For example:

- [doit interfering with the command history of pdb](https://github.com/scipy/scipy/issues/16452)
- Emoji rendering issues in the Windows CI environment [[gh-18046](https://github.com/scipy/scipy/issues/18046)]
- Inability of `dev.py` to resolve Homebrew-installed Python [[gh-18998](https://github.com/scipy/scipy/issues/18998)]

Many of these problems stem from `doit`, which is a dependency of `dev.py`.
One of the advantages of `spin` is that it has fewer dependencies, which in turn means fewer points of failure.
While this trade-off sacrifices some of the output’s polish, it greatly improves the reliability of the development process for SciPy.

> **Why not just fix `dev.py`?**
> 
> You might wonder, "Why not just update `dev.py` to fix these issues? Why switch to a new third-party tool like `spin`?"

Reviewer: ```suggestion
You might wonder, "Why not just update `dev.py` to fix these issues? Why switch to a new third-party tool like `spin`?"
```

Removing since it feels redundant.

One of the key reasons is that **`spin` is already widely used by other scientific computing projects** such as `numpy`, `scikit-image`, `solcore5`, `skmisc`, `PyFVS`, `sktree`, `dipy`, and many more (refer to [scientific-python/spin#62](https://github.com/scientific-python/spin/issues/62)).
This means improvements made to `spin` will automatically benefit SciPy.

> Additionally, any limitations we encounter in `spin` will be addressed through Pull Requests, making it a two-way connection: SciPy benefits from other projects (via `spin`), and those projects benefit from improvements made by SciPy.

Reviewer: ```suggestion
Additionally, any limitations we encounter in `spin` will be addressed through upstream contributions to the project, making it a two-way connection: SciPy benefits from other projects contributing to `spin`, and those projects benefit from improvements made by the SciPy team to `spin`.
```

**Fixing a linker path issue on macOS**

While building SciPy with `spin`, we encountered an issue on macOS where the linker path for shared libraries became corrupted due to the `--destdir` specification in the `meson install` step.
This issue arose because `spin` sets `/usr` or `C:\` as the default value for `--prefix`, but it also used `--destdir` during installation.
`--destdir` is typically used for packaging, not for final installations.
To resolve this, we removed the `--destdir` usage and set the `--prefix` directly to the installation directory.
The fix was submitted as a PR to `spin` [[scientific-python/spin#257]](https://github.com/scientific-python/spin/pull/257), ensuring smoother installation behavior for SciPy on macOS.

### Current Status of `spin` in SciPy

We are close to merging `DEV: use spin (prototype)` [[gh-21674](https://github.com/scipy/scipy/pull/21674)].
We’re waiting for the PR [scientific-python/spin#257](https://github.com/scientific-python/spin/pull/257) to be merged, after which we’ll be ready to fully integrate `spin` into SciPy.
Our work has been fully tested in the SciPy CI system, ensuring its stability.

> Here’s a quick comparison of the SciPy build experience with `dev.py` versus `spin`:

Reviewer: The videos don't seem to rendering in the preview: https://labs-git-fork-czgdp1807-feature-intel-oneapi-spin-quansight.vercel.app/blog/intel-oneapi-msvc-spin-scipy

Author: I see, I can try https://docs.asciinema.org/manual/server/embedding/#preview-image-link for sharing the videos. `asciinema` will provide an image. If someone clicks it then the videos plays on asciinema website. That should solve the problems showing up due to embedding the player using `script` tag. Let me know if it would work.

Reviewer 2: That seems like a pragmatic solution - try it I'd say.

Reviewer: I pushed up a commit to do this (please let me know if you're not comfortable with this, and I'll keep it mind for the future), and unblock merging this PR. :)

Author: No worries. Thanks for doing it. :D.

**SciPy Build Experience with `dev.py`**

<script src="https://asciinema.org/a/fn1JP6ZNPpw6rKh8oAusk34Xr.js" id="asciicast-fn1JP6ZNPpw6rKh8oAusk34Xr" async="true"></script>

This video shows building SciPy with `dev.py`.

**SciPy Build Experience with `spin`**

<script src="https://asciinema.org/a/TdMa1hSSy3MTPnFcYKHvP0pvm.js" id="asciicast-TdMa1hSSy3MTPnFcYKHvP0pvm" async="true"></script>

This video shows building SciPy with `spin`.

As you can see there is no difference between the look and feel of `python dev.py build` and `spin build`.

> It works exactly the same. However, the issues which were there with `dev.py` are not observed with `spin`.

Reviewer: ```suggestion
It works exactly the same. Moreover, the issues which were there with `dev.py` are not observed with `spin`.
```

Take a look at the following videos,

**`spin` keeps pdb command history intact while `dev.py` interferes with it [[gh-16452]](https://github.com/scipy/scipy/issues/16452)**

**With `dev.py`**

<script src="https://asciinema.org/a/jtTP2xty4popJ1GVL0PZQBsvz.js" id="asciicast-jtTP2xty4popJ1GVL0PZQBsvz" async="true"></script>

This video shows `dev.py` interfering with `pdb` command history.

**With `spin`**

<script src="https://asciinema.org/a/uOT0cp0WJSKnDZPmRqdm0vREC.js" id="asciicast-uOT0cp0WJSKnDZPmRqdm0vREC" async="true"></script>

This video shows `spin` keeps `pdb` command history intact.

> As you can see `spin` usage doesn't interfere with PDB's history. This issue was there with `dev.py`.

Reviewer: ```suggestion
As you can see `spin` usage doesn't interfere with PDB's history, which was an issue with `dev.py`.
```

On Linux, the issue was already fixed. However, there were still some challenges when using `ifx`, such as the failure of the `test_equal_bounds` test in `scipy/optimize/tests/test_optimize.py` due to floating-point discrepancies.
To address this, we increased tolerances for several tests to accommodate the slightly reduced accuracy/precision when using `ifx`.
This was all handled in the PR, `BUG/CI: Compile and run tests with ifx + MKL on Linux` [[gh-21173]](https://github.com/scipy/scipy/pull/21173), where I also added a CI job to test the combination of `gcc` + `g++` + `ifx` + MKL on Linux, marking another important step toward full Intel oneAPI support for SciPy.

The final milestone for Intel oneAPI support was the CI job for testing SciPy with `icx`, `icpx`, `ifx`, and MKL [[gh-21254]](https://github.com/scipy/scipy/pull/21254).
In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux.

However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`.

> Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work, we decided to pause the effort on Windows and focus on more scoped tasks.

Reviewer 2: ```suggestion
Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work and the SciPy team is planning to rewrite ARPACK completely in a language other than FORTRAN, we decided to pause the effort on Windows and focus on more scoped tasks.
```

---

> title: 'Enhancing Developer Experience at SciPy - Intel oneAPI/MSVC Support and Migrating to `spin`'

Author: Should I remove, "\`" from "`spin`"?

published: December 19, 2024
authors: [gagandeep-singh]
description: 'Highlights the work done to improve developer experience at SciPy, specifically on supporting Intel oneAPI/MSVC and spin'
category: [Developer workflows]
featuredImage:
  src: /posts/intel-oneapi-msvc-spin-scipy/scipy_logo_img_featured.png
  alt: 'SciPy logo'
hero:
  imageSrc: /posts/intel-oneapi-msvc-spin-scipy/scipy_logo_img_hero.png
  imageAlt: 'SciPy logo'
---

In this blog post, we will be sharing the work we did to enhance developer experience at SciPy. We worked on two major things,

1. Intel oneAPI and MSVC Support in SciPy
2. Moving away from `dev.py` and migrating to `spin`.

> First, let me share the motivation behind these.

Reviewer 2: It would be better I think to share the general motivation in fewer words. In-depth paragraphs about both topics before repeating part of the in-depth content in the separate sections on OneAPI and `spin` feels a little off.

> Intel offers a complete toolkit for high performance computing (HPC), numerical computing use cases. It goes by the name of [oneAPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/overview.html). It provides, ICX (for C), ICPX (for C++), IFX (for Fortran) and Math Kernal Library (MKL for optimized math routines like BLAS, LAPACK, fast FFT, etc.) [[1]](https://en.wikipedia.org/wiki/Math_Kernel_Library). Since, SciPy caters to HPC and numerical computing use cases, so supporting building SciPy with Intel oneAPI is a natural requirement. In fact, users reported segmentation fault in `arpack` with `ifx` [[gh-20728]](https://github.com/scipy/scipy/issues/20728) and `scipy.linalg.solve` crash when used with Intel oneAPI on Windows 11 [[gh-21997]](https://github.com/scipy/scipy/issues/21997) are two such examples. Therefore we decided to work on this to enhance the experience of SciPy users who use Intel oneAPI as their toolkit. Regarding MSVC, since it is the de-facto compiler for C/C++ on Windows. So adding support for it is again a natural requirement and also a must.

Reviewer 2: - I'd use "MKL (Math Kernel Library, ...)" rather than the other way around, since you're also using the abbreviations for ICX & co
- MSVC: we didn't actually add support for it though, so the rationale here is a little off. Try to rephrase to what we actually aimed to do.

Author: I think MSVC wasn't tested on CI before our work, right? On Windows we used clang for CI testing. So may be a better line can be, "Since it is the de-facto compiler for C/C++ on Windows, we added CI jobs for testing SciPy compilation with MSVC"

> Now let's see why we decided to spin from `dev.py` to `spin` (pun inteded). `dev.py` (or `do.py` [[2]](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was added to improve the experience for SciPy developers. Basically, it offers better documentation for the various commands needed (like, `build`, `test`, `shell`) while working on SciPy codebase. It has better looking CLI outputs. However, with time several issues with `dev.py` were reported. For example, in some cases, doit based dev interface interferes with pdb command history [[gh-16452]](https://github.com/scipy/scipy/issues/16452), `dev.py` giving problems in a Windows CI environment because of emojis usage in dev.py [[gh-18046]](https://github.com/scipy/scipy/issues/18046), and dev.py unable to resolve homebrew installed Python correctly [[gh-18998]](https://github.com/scipy/scipy/issues/18998). `spin` easily fixes this issues. [[3]](https://github.com/scipy/scipy/pull/21674), especially those issues arising due to `doit`. `dev.py` depends on `doit` but `spin` doesn't. Also, since `spin` is used by many other projects, so all the improvements in `spin` due to the issues reported by their maintainers will directly benefit SciPy.

Reviewer 2: I'd say the long-term maintenance load sharing and the familiarity for contributors coming from `numpy` or other project are the main motivations. The paper cuts & bugs are very minor.

Reviewer 2: Also, it'd be good to say somehting like "`dev.py` was a major improvement in development workflow, and came before `spin`. `spin` took its inspiration [in large part](https://github.com/scientific-python/spin/?tab=readme-ov-file#history) from `dev.py`, and generalized its interface to work with more projects without customization as heavy as `dev.py` has for SciPy".

#### Intel oneAPI Support in SciPy

> Intel offers oneAPI - a complete toolkit for HPC use cases. It has ICX (for C), ICPX (for C++), IFX (for Fortran) and MKL (an OpenBLAS alternative). We targeted two operating systems, Windows and Linux. `CI: adding a Windows CI job with MSVC + MKL + Intel Fortran (ifx)` [[gh-20878](https://github.com/scipy/scipy/issues/20878)] was the starting point. Solving this issue meant, avoiding regressions with MSVC and Intel oneAPI in SciPy. This issue also links [gh-20728](https://github.com/scipy/scipy/issues/20728). The author of this issue is building SciPy with `icx`, `icpx`, `ifx` and using MKL for BLAS and LAPACK. So a complete usage of Intel oneAPI. In addition, failure to build scipy with msvc [[gh-20860]](https://github.com/scipy/scipy/issues/20860) is related to build failure of SciPy with MSVC. We started off by fixing these two.

Reviewer 2: This "It has ICX ..." is repeated from a few paragraphs up. I actually suggest moving that content here.

One major issue with MSVC is that its support for handling arrays on stack (with runtime sizes) is limited. Expressions like, `std::complex<double> cwrk[n];` fail to compile with MSVC unless `n` is a constant. So we did runtime allocation on heap (using `new`) and performed `delete` at the end to free memory. This is a manual approach of handling memory, however this is the only solution which works with MSVC. It was the only option that we went ahead with. On a side note, Clang due to LLVM backend doesn't have this limitation. After fixing this issue we added a CI job with MSVC + `ifx` + OpenBLAS combination. This helped in avoiding future regression with MSVC build of SciPy. It was also a first step towards supporting Intel oneAPI toolkit entirely.

> Now coming on to the `arpack` issue. Fixing this issue required supporting building SciPy with Intel oneAPI. On Linux the `arpack` issue was already resolved. However, there were some other issues with `ifx`. For example, `test_equal_bounds` test in `scipy/optimize/tests/test_optimize.py` failed due to floating point issues with `ifx`. Also tolerances for other tests had to be increased (i.e., slight lesser accuracy/precision in order to make things with `ifx`). All of this was done in `BUG/CI: Compile and run tests with `ifx`+`MKL` on Linux` [[gh-21173]](https://github.com/scipy/scipy/pull/21173). I also added a CI job in this PR with `gcc` + `g++` +`ifx` + `MKL` combination. A second steps towards supporting Intel oneAPI with SciPy.

Reviewer 2: You may want to link to https://github.com/opencollab/arpack-ng, most readers won't know what ARPACK is.

> The final checkpoint for Intel oneAPI was `CI: Test icx + icpx + ifx + MKL build of SciPy` [[gh-21254]](https://github.com/scipy/scipy/pull/21254). Here we replaced `gcc` with `icx` and `g++` with `icpx` and TADA, it worked on Linux. Regarding Windows, we are still at `MSVC` + `ifx` + `OpenBLAS` combination because of the `arpack` import error. We tried several ways to fix it but all resulted in a dead end. We tried to build NumPy with `icx`, `icpx` but `icx` is unable to compile NumPy code as is. Some workarounds had to be made. However, still it didn't work for us.

Reviewer 2: The last two sentences are not a very good ending. Try to say why we stopped (e.g., "the effort wasn't justified by ..."), otherwise it leaves the reader wondering.

#### Moving away from `dev.py` and migrating to `spin`

Before diving into the details of this work, I would first share some information (for context) related to `meson`. Basically,`meson` builds a project in three steps,

1. First step is that it calls `meson setup` which does configuration for building the project. If you are a CMake user then its just like calling, `cmake .` in a project.
2. Second step involves calling `meson compile` which executes the compiler commands and creates libraries. Again if you are `cmake` + `make` user then its like calling, `make` or `make -j8`.
3. Third step is finally calling, `meson install` which installs the libraries, source files (for python) in the installation directory.

Now it would be easy to understand our work below.

> We decided to shift to `spin` because of the several issues reported in `dev.py`. For example, [doit interfering with command history of pdb] [[gh-16452]](https://github.com/scipy/scipy/issues/16452), usage of emojis giving issues in a Windows CI environment [[gh-18046]](https://github.com/scipy/scipy/issues/18046), and dev.py unable to resolve homebrew installed Python [[gh-18998]](https://github.com/scipy/scipy/issues/18998). A bunch of those are due to `doit` being a dependency of `dev.py`. One nice feature of `spin` is it has lesser dependencies. This means, lesser bugs because more dependencies means more points of failure. The tradeoff here is to compromise looks and feel of the output a little bit to avoid problems while developing SciPy.

Reviewer 2: repeating grammar issue, so commenting: it's "fewer bugs" rather than "lesser bugs" (check for "lesser" in the whole post please)

<script src="https://asciinema.org/a/FJzWyQun71g796n5kFI1l6dxv.js" id="asciicast-FJzWyQun71g796n5kFI1l6dxv" async="true"></script>

As you can see there is no difference between look and feel of `python dev.py build` and `spin build`. It works exactly the same. However, the issues which were there with `dev.py` are not observed with `spin`. Take a look at the following videos,

**`spin` keeps pdb command history intact while `dev.py` interferes with it [[gh-16452]](https://github.com/scipy/scipy/issues/16452)**

**With `dev.py`**

<script src="https://asciinema.org/a/shvsE0PfbGMsviuthoKfIQlI6.js" id="asciicast-shvsE0PfbGMsviuthoKfIQlI6" async="true"></script>

**With `spin`**

<script src="https://asciinema.org/a/wcwYTiPQzey2uYt1082w63Rgz.js" id="asciicast-wcwYTiPQzey2uYt1082w63Rgz" async="true"></script>

As you can see `spin` usage doesn't interfere with PDB's history. This issue was there with `dev.py`.

**Using homebrew installed Python to build SciPy [[gh-18998]](https://github.com/scipy/scipy/issues/18998) - `dev.py` vs `spin`**

<script src="https://asciinema.org/a/aZstdaf6B6nlkW9JFccT7z7Dr.js" id="asciicast-aZstdaf6B6nlkW9JFccT7z7Dr" async="true"></script>

> The above video showcases how `spin` works fine with Homebrew installed Python 3.10. The build completes. However, `dev.py` fails because it is unable to resolve paths correctly.

Reviewer 2: It would be good to add acknowledgements - this was done with NASA funding. Can you please copy the phrasing at the end of Albert's recent blog post for this?
